File: src/obsidian_mcp/links.py
"""Wikilink insertion and backlink/forward-link scanning."""

from __future__ import annotations

import re

from mcp.server.fastmcp import FastMCP, Context

from obsidian_mcp.errors import NoteNotFoundError
from obsidian_mcp.vault import Vault

WIKILINK_RE = re.compile(r'\[\[([^\[\]]+)\]\]')


def insert_wikilink(
    vault: Vault, source_path: str, target_title: str, alias: str | None = None
) -> str:
    note_path = vault.resolve(source_path)
    if alias:
        link = f"[[{target_title}|{alias}]]"
    else:
        link = f"[[{target_title}]]"

    content = note_path.read_text(encoding="utf-8")
    if content and not content.endswith("\n"):
        content += "\n"
    content += link + "\n"
    note_path.write_text(content, encoding="utf-8")
    return str(note_path.relative_to(vault.root))


def get_backlinks(vault: Vault, path: str) -> list[str]:
    target_path = vault.resolve(path)
    stem = target_path.stem
    rel = str(target_path.relative_to(vault.root))
    rel_no_ext = rel[:-3] if rel.endswith(".md") else rel

    results: list[str] = []
    for md_file in vault.all_markdown_files():
        if md_file == target_path:
            continue
        content = md_file.read_text(encoding="utf-8")
        for match in WIKILINK_RE.finditer(content):
            link_target = match.group(1).split("|")[0].strip()
            if link_target == stem or link_target == rel_no_ext:
                results.append(str(md_file.relative_to(vault.root)))
                break
    return results


def get_forward_links(vault: Vault, path: str) -> list[str]:
    note_path = vault.resolve(path)
    content = note_path.read_text(encoding="utf-8")
    links = []
    for match in WIKILINK_RE.finditer(content):
        target = match.group(1).split("|")[0].strip()
        links.append(target)
    return links


def register_tools(mcp: FastMCP) -> None:
    """Register link tools with the MCP server."""
    from mcp.server.fastmcp.exceptions import ToolError

    @mcp.tool()
    async def link_insert_wikilink(
        source_path: str, target_title: str, alias: str | None = None, ctx: Context = None
    ) -> str:
        """Append a wikilink to a note."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return insert_wikilink(vault, source_path, target_title, alias)
        except NoteNotFoundError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def link_get_backlinks(path: str, ctx: Context = None) -> list[str]:
        """Find all notes that link to the given note."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return get_backlinks(vault, path)
        except NoteNotFoundError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

    @mcp.tool()
    async def link_get_forward_links(path: str, ctx: Context = None) -> list[str]:
        """Return all wikilinks in a note."""
        vault: Vault = ctx.request_context.lifespan_context.vault
        try:
            return get_forward_links(vault, path)
        except NoteNotFoundError as e:
            raise ToolError(str(e))
        except OSError as e:
            raise ToolError(f"I/O error: {e}")

File: rust/src/links.rs
use std::fs;
use regex::Regex;

use crate::errors::Result;
use crate::vault::Vault;

/// Regex that matches Obsidian wikilinks: [[target]] or [[target|alias]]
fn wikilink_re() -> Regex {
    Regex::new(r"\[\[([^\[\]]+)\]\]").unwrap()
}

/// Append a wikilink to the end of a note. Returns the note's relative path.
pub fn insert_wikilink(
    vault: &Vault,
    source_path: &str,
    target_title: &str,
    alias: Option<&str>,
) -> Result<String> {
    let note_path = vault.resolve(source_path)?;
    let link = match alias {
        Some(a) => format!("[[{}|{}]]", target_title, a),
        None => format!("[[{}]]", target_title),
    };

    let content = fs::read_to_string(&note_path)?;
    let mut new_content = content;
    if !new_content.is_empty() && !new_content.ends_with('\n') {
        new_content.push('\n');
    }
    new_content.push_str(&link);
    new_content.push('\n');

    fs::write(&note_path, new_content.as_bytes())?;
    Ok(vault.relative(&note_path))
}

/// Find all notes that contain a wikilink pointing to the given note.
/// Returns relative paths of files that link to it.
pub fn get_backlinks(vault: &Vault, path: &str) -> Result<Vec<String>> {
    let target_path = vault.resolve(path)?;
    let stem = target_path
        .file_stem()
        .map(|s| s.to_string_lossy().to_string())
        .unwrap_or_default();
    let rel = vault.relative(&target_path);
    let rel_no_ext = if rel.ends_with(".md") {
        rel[..rel.len() - 3].to_string()
    } else {
        rel.clone()
    };

    let re = wikilink_re();
    let mut results: Vec<String> = Vec::new();

    for md_file in vault.all_markdown_files() {
        if md_file == target_path {
            continue;
        }
        let content = match fs::read_to_string(&md_file) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let mut linked = false;
        for cap in re.captures_iter(&content) {
            let link_target = cap[1].split('|').next().unwrap_or("").trim();
            if link_target == stem || link_target == rel_no_ext {
                linked = true;
                break;
            }
        }
        if linked {
            results.push(vault.relative(&md_file));
        }
    }
    Ok(results)
}

/// Return all wikilink targets referenced from a note.
pub fn get_forward_links(vault: &Vault, path: &str) -> Result<Vec<String>> {
    let note_path = vault.resolve(path)?;
    let content = fs::read_to_string(&note_path)?;
    let re = wikilink_re();
    let links: Vec<String> = re
        .captures_iter(&content)
        .map(|cap| cap[1].split('|').next().unwrap_or("").trim().to_string())
        .collect();
    Ok(links)
}
